### PR TITLE
codeintel: Reduce memory when calculating commit graph

### PR DESCRIPTION
**Pull request by efritz**
This is an effort to reduce memory pressure as described in https://github.com/sourcegraph/sourcegraph/issues/16030.

## What changed?

We join lsif_nearest_uploads with lsif_uploads to bring back the root and indexer values of each upload to determine which uploads can "shadow" another's visibility on the git commit graph. We bring these fields back with full information, although we only need to determine if they are the same or not.

We instead bring back an md5 hash of those two fields. If the root and indexer fields are the same, the hash will also be the same. And, with acceptable likelihood, two different hashes have a different root or indexer.

This saves us many duplicate copies of the same strings in memory, which was causing a lot of pressure.

## Results

**TL;DR:** This change will make the graph update procedure 2.5x slower, but will take only 1/4 of the memory.

**Before:**
```
$ go test -run='^BenchmarkCalculateVisibleUploads$' -bench=. -benchmem -v
goos: darwin
goarch: amd64
pkg: github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/stores/dbstore
BenchmarkCalculateVisibleUploads
BenchmarkCalculateVisibleUploads-16    	       1	3466152664 ns/op	415819920 B/op	   63756 allocs/op
PASS
ok  	github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/stores/dbstore	3.726s

```

**After:**
```
$ go test -run='^BenchmarkCalculateVisibleUploads$' -bench=. -benchmem -v
goos: darwin
goarch: amd64
pkg: github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/stores/dbstore
BenchmarkCalculateVisibleUploads
BenchmarkCalculateVisibleUploads-16    	       1	8871662674 ns/op	107866832 B/op	   63830 allocs/op
PASS
ok  	github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/stores/dbstore	9.182s
```

<details>
<summary>Benchmark test (run on main@HEAD)</summary>

```go
func BenchmarkCalculateVisibleUploads(b *testing.B) {
	var (
		numCommits = 5000
		numUploads = 2000
		numRoots   = 500
	)

	var commits []string
	for i := 0; i < numCommits; i++ {
		commits = append(commits, fmt.Sprintf("%40d", i))
	}

	graph := map[string][]string{}
	for i, commit := range commits {
		if i == 0 {
			graph[commit] = nil
		} else {
			graph[commit] = commits[i-1 : i]
		}
	}

	var roots []string
	for i := 0; i < numRoots; i++ {
		roots = append(roots, fmt.Sprintf("sub%d/", i))
	}

	uploads := map[string][]UploadMeta{}
	for i := 0; i < numUploads; i++ {
		x := fmt.Sprintf("%40d", int((float64(i*numUploads) / float64(numCommits))))
		uploads[x] = append(uploads[x], UploadMeta{
			UploadID: i,
			Root:     roots[i%numRoots],
			Indexer:  "lsif-go",
		})
	}

	b.ResetTimer()
	b.ReportAllocs()

	for i := 0; i < b.N; i++ {
		_, _ = calculateVisibleUploads(graph, uploads)
	}
}
```
</details>

### CODE REVIEW

**Comment by aisbaa** [comment]
I'm ready to test this, let me know when image is ready.

File: enterprise/internal/codeintel/stores/dbstore/commit_graph.go [hunk 1]
```diff
@@ -144,27 +125,30 @@ func calculateVisibleUploads(graph map[string][]string, uploads map[string][]Upl
...
-						Overwritten:     true,
-					})
-				}
-				u[i] = upload
-				return u
-			}
+		candidateSharedFieldToken := commitGraphView.Tokens[candidate.UploadID]
+		if sharedFieldToken != candidateSharedFieldToken {
+			continue
+		}
```
Comment (Strum355) [inline enterprise/internal/codeintel/stores/dbstore/commit_graph.go:135]: Im curious on your thoughts of using a bloom filter here instead of a map?

Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/commit_graph.go:135]: Could you elaborate? My first instinct is that it would tell us there _is_ a conflict but wouldn't identify which element (which we need to mark with the overwritten flag).

Comment (Strum355) [reply enterprise/internal/codeintel/stores/dbstore/commit_graph.go:135]: `sharedFieldToken` and `candidateSharedFieldToken` are the only two variables that are retrieved from the commitGraphView.Tokens map, and theyre only used for immediate comparison. I was thinking we could use a bloom filter instead of keeping a full map of all upload IDs. Havent used bloom filters to enough of a degree to know if it can be used for this kv use-case rather than as a set, but a quick google suggests that it could be?

**Review by Strum355 — APPROVED** [review]
